ansible_base/authentication/utils/authenticator_map.py

```diff
@@ -18,12 +30,166 @@ def has_expansion(value: Optional[str]) -> bool:
         return _("Expansion only supports the format {% for_attr_value(attribute) %}")
+
+
+def expand_syntax(attributes: dict, map: AuthenticatorMap) -> list[Dict[str, str]]:
+    """
+    Given attributes and a map, look for the fields that can be expanded and do the expansion
+    If no fields require expansion, do nothing
+
+    returns:
+    list of named tuples of expanded value
+    """
+    expanded_strings = {}
+    for field in _EXPANSION_FIELDS:
+        field_value = getattr(map, field, None)
+
+        if field_value is None:
+            # If this particular field is None we can move on.
+            # This might be true for the team value of an authenticator map type of organization for example.
+            continue
+
+        # Get a list of all of the attributes specified in {% for_attr_value(something) %}
+        # In the above example, we would get a list like ['something'].
+        # If there were more than one in a given string we would have an array like ['something', 'else']
+        attrs_used = _EXPANSION_RE.findall(field_value)
+
+        if has_expansion(field_value) and attrs_used == []:
+            # This field had an invalid expansion so we can't expand anything, just return []
+            logger.info(f"Authenticator Map {map.name} on {map.authenticator.name} has a bad expansion in {field} unable to process map ({field_value})")
+            continue
+
+        # Make sure that the attributes we got contain all the ones we want to expand
+        if missing_attributes := set(attrs_used).difference(set(attributes.keys())):
+            logger.info(
+                f"Authenticator Map {map.name} on {map.authenticator.name} tried to expand attribute(s) "
+                f"{', '.join(missing_attributes)} but they were not in the users attributes!"
+            )
+            continue
+
+        # Normalize and check the attribute values we are going to use
+        normalized_attributes = {}
+        had_issues = False
+        for attr_name in attrs_used:
+            # Make sure the attribute is in the list of attributes we have.
+            normalized_attributes[attr_name] = attributes.get(attr_name, [])
+
+            # Make this a list if its not already a list. We can't do this as a typecast because list(dict) returns a list of keys
+            if type(normalized_attributes[attr_name]) is not list:
+                normalized_attributes[attr_name] = [normalized_attributes[attr_name]]
+
+            if len(normalized_attributes) == 0:
+                logger.info(
+                    f"Authenticator Map {map.name} on {map.authenticator.name} tried to expand attribute {attr_name} "
+                    f"but there were not values in that attribute, ignoring"
+                )
+                had_issues = True
+
+            # If any of the values are not a string then we can't expand it
+            for value in normalized_attributes[attr_name]:
```

Comment:
I'd suggest something like this instead:

`non_strings = [value for value in normalized_atttributes[attr_name] if not isinstance(value, str)]`
`for non_string in non_strings:`
    `...do the logging...`

to log all invalid attributes rather than just the first that the `break` would result in.